### PR TITLE
dev/core#3035 Add New Extension: Display all stable extensions (not just reviewed ones)

### PR DESCRIPTION
**Pull request by mlutfy**
Overview
----------------------------------------

Gitlab: https://lab.civicrm.org/dev/core/-/issues/3035

This proposes that, as of CiviCRM 5.79 (master), the in-app Extension Browser (add new) displays all stable extensions, not just the ~~stable~~ reviewed ones.

There are also a few small changes:

- Extensions tab:
  - Renamed the header "Name" to "Extension" (to be the same as on the "Add New" tab)
- Add New tab:
  - Removed "Type" column (was always "module")
  - Added "Stability" column (alpha, beta, stable, reviewed)
  - "Download" action renamed to "Install" ("download" is a technicality, the actual action is "install")
  - Removed jsortable, so that the sort order is more predictable (same as on civicrm.org) - I'm on the fence on this, maybe people want to sort alphabetically if they don't know that their browser has a search function?
- When displaying the details (uncollapse the fieldset)
  - Added row for the number of installs (usage)
  - Re-order a few items in a way that felt more intuitive to me

It also fixes the issue reported in dev/core#3035, which is that non-reviewed extensions did not display that an upgrade was available.

Before
----------------------------------------

image

After
----------------------------------------

(the visual is a bit different because this is from Standalone, not Drupal7)

image

and also non-reviewed extensions have upgrade notifications:

image

Technical Details
----------------------------------------

I did some changes in the `extdir` module on civicrm.org, so that starting from CiviCRM 5.79, it lists all extensions by default, but does not change the behaviour for older versions of CiviCRM (because they do not display the "reviewed" status, scary messages, etc). Here are the changes, and they are live: https://lab.civicrm.org/marketing/civicrm-website/-/commit/c6ccc013a98ffb385db7d8e714bbe34f5f9e295c

The stability/popularity sorting is done by civicrm.org. Because of how things are stored in XML, it's a bit of a mess to handle client-side.

**Important:** the extdir is significantly slower and can sometimes timeout (more than 10 seconds?). It gets cached by civicrm.org, but still, that will definitely need to be improved.

### CODE REVIEW

**Comment by stesi561** [comment]
> displays all stable extensions, not just the stable ones.
what does this mean?

**Comment by MegaphoneJon** [comment]
IMO unreviewed extensions shouldn't appear unless someone checks a box indicating they want them.  It's too easy to screw up your install by downloading a half-baked extension.  Just today I downloaded an extension that's in the directory that takes out your site because a) it's a WIP, b) it requires a change to `civicrm.settings.php` to work.

Re: the caching - I've hit this issue many times myself.  Perhaps instead of caching when requested, we can generate with a cron job? Or is generated as part of the script that runs when a new (extension) release is tagged?

Re: the XML - I assume that the source data is from Drupal services?  Or a View with XML display?  Could we also generate a JSON version, with newer Civi versions consuming the JSON version?

Sorry - I know this is massive scope creep, but the first issue is IMO a blocker.

If we *really* wanted to scope creep - there's been discussion on MM about SearchKit parsing JSON data (a pre-requisite for having a FormBuilder "results" search).  If we had that, we could just use SK for the Extensions screen.

**Comment by MegaphoneJon** [comment]
> > displays all stable extensions, not just the stable ones.
> > what does this mean?

There are a lot of extensions in the extension directory that haven't been approved for in-app use because they haven't gone through a review process.  E.g. [Discount Contributions](https://civicrm.org/extensions/discount-contributions).  This PR proposes to expose those all to the in-app installer/browser.

**Comment by stesi561** [comment]
> IMO unreviewed extensions shouldn't appear unless someone checks a box indicating they want them. It's too easy to screw up your install by downloading a half-baked extension.

Agree - or even a fully baked one that has been put on the shelf and not used, PHP and CiviCRM versions move on and now we have a fatal error.

My assumption would be people using this UI should either be warned in very very large SCARY red letters that - installing an extension can break you site in a way you can't fix unless you have access to the server to delete the extension. Or we should not allow this workflow.

**Comment by twomice** [comment]
> IMO unreviewed extensions shouldn't appear unless someone checks a box indicating they want them.

I pretty strongly agree. 

The proposed change is not just a change in functionality, but a change to the _trust model_, from the current "I can expect that any in-app extension has passed some kind of quality standard" to "I need to think a lot more carefully about which in-app extensions to install."

That merits requiring an explicit opt-in of some kind; otherwise, I (the non-technical site admin) am very likely getting elevated risk without realizing it.

> displays all stable extensions, not just the stable ones.

I think this is a typo. @mlutfy should this not say, "displays all stable extensions, not just those that have passed review" ?

**Comment by mlutfy** [comment]
> "displays all stable extensions, not just the stable ones."

Fixed typo in PR summary to "not just the reviewed ones"

> "IMO unreviewed extensions shouldn't appear unless someone checks a box indicating they want them."

Publishing an extension in the directory is already pretty difficult.

And that's kind of the basis of the extension directory: publish it, and people might install it.

We should add more hoops. But we can clarify the in-app messaging. If extensions aren't stable, we should change their status.

> "My assumption would be people using this UI should either be warned in very very large SCARY red letters that - installing an extension can break you site in a way you can't fix unless you have access to the server to delete the extension. Or we should not allow this workflow."

So we're developers and we think towards the worst case scenario, but I think we have to work with a few assumptions:

- Most admins probably don't want to do FTP/ssh, but at some point, either they or someone they work with knows how to do so. If they break the system, they can still fix it the old fashioned way.
- People usually report bugs on issue queues. Maybe 10-20% of admins know how to do that, but eventually someone does. Then we can change the stability back to "beta" or "unsupported" and it won't show in-app (and we do so from time to time). We can be more aggressive about updating the status, let's fix the actual problem, because I might install an extension manually, and it will crash my site, and I'll still be annoyed, even if I know how to fix it.
- Admins might be poking around on some dev environment, where they can easily spin up a dev site, install extensions, crash, and discard their environment. Admittedly, it's not very common, but it increasingly is (ex: Bitnami)
- This problem of PHP compatibility and CiviCRM core versions also applies to reviewed extensions.

And those problems are likely, but not very frequent. Whereas there are regular complaints about 1) not being able to install a non-reviewed extension, and 2) not seeing that upgrades are available. Problem (2) could have been solved otherwise, but I think we should fix (1), which also fixes (2).

There is also a marketing aspect: say I heard that CiviCRM now has Zapier integration. That extension currently has 6 active sites using it. Should I use it? Can I try it out on a dev site? Why do I need to hire someone to install it?

> "very very large SCARY red letters"

Yes, we should help admins make a better decision. One thing that concerns me about this page, is that it still has a fair amount of clutter, but doesn't really communicate the important bits, like more info about the extension (the "description" is too short).

> Re: the caching - I've hit this issue many times myself. Perhaps instead of caching when requested, we can generate with a cron job? Or is generated as part of the script that runs when a new (extension) release is tagged?

Probably - but I guess I'll circle back if we get some consensus on this PR.  It's a custom  Drupal module written some time ago by someone else. I ported it to Drupal8+. I linked the commit of my changes in the PR summary.

There is also a standalone PHP application running on top of it, which is supposed to do its own caching, but while I was testing, that application seemed to be adding latency as well. And it's still not ported to PHP8:

- https://lab.civicrm.org/infra/extdir/-/tree/master
- https://lab.civicrm.org/infra/ops/-/issues/1010

> Re: the XML [...] Could we also generate a JSON version, with newer Civi versions consuming the JSON version?

Do you consider that a blocker for this PR?

**Comment by stesi561** [comment]
+1 for making it easier to see important information.  

Sorry perhaps I misunderstood the scope of what would be showing.

> Maybe 10-20% of admins know how to do that, but eventually someone does.
I guess you could argue that people who know how to do it are more likely to lodge a issue.  In which case I think if being more aggressive about reverting.

> This problem of PHP compatibility and CiviCRM core versions also applies to reviewed extensions.
Agree so perhaps the wider discussion about the usability of the page can be separate from this.

**Comment by guyiac** [comment]
I agree with @twomice that this is a change in the trust model. To me, it makes a lot of sense that you can load any extension you want from anywhere with back end access, that there are lots of extensions on civicrm.org that meet some minimum standard, and that extensions in app have met some basic quality standard (review).  That was a conscious choice to do that that way with a lot of input from a good number of people back in Fort Collins, and I think it's still valid today for a lot of the reasons above.  I would like it though if there was a way that non-reviewed extensions showed available updates.

I also think we need to consider unintended consequences. Right now some devs won't get an extension reviewed because they don't want to handle requests for free support, but they will make the extension available on civicrm.org. If these started showing up in app, some devs might not even make them available on civicrm.org, making less extensions easily available.

I don't think the risks and drawbacks associated with this change far outweigh the benefits. However, all for cleaning up the interface and don't have any strong opinions there - everything you have chosen there makes sense to me.

**Comment by ufundo** [comment]
> I guess you could argue that people who know how to do it are more likely to lodge a issue.

I am just one user - but I think it would be nice for the links to the extension issue queue to be more prominent (and maybe have a nice indicative logo?)

I think that could help: a) people see if there are issues relevant to their use case before they install; b) encourage people to report issues they hit; c) generally channel people through to "the community"

I can see that it might be frustrating to extension developers if this feels like free support request -- but if issues are out in the open on trackers then 1) other people might fix them; 2) offers of funding might arise.

Many eyes is what open source is about, right?

**Comment by ufundo** [comment]
Personally I'm less worried about the trust model thing, because I've always locked down in-app install on production sites. As @mlutfy says, there's reviewed extensions that go bang too, so personally I don't think the reviewedness is a good enough guard.

Also +1 for being able to use in disposable test/demo environments.

**Comment by mlutfy** [comment]
Excellent feedback so far. I sincerely appreciate it. I tackled this issue as a usability issue for new admins, and there are important things not to neglect with regards to the trust model, and with expectations towards developers/support.

> [Paraphrasing]  Developers might get too many support requests

I think that would be a good problem. We have to humbly acknowledge that the community of DIY CRM enthusiasts is not what it was. Like many other Free Software communities, we stagnated, while the overall CRM market doubles every few years. As a result, some FOSS projects went "enterprise", such as Drupal, or some just keep to their niche (that's us).  I'm not against "enterprise" (SaaS & Spark?), but we also have to empower the DIY folks if we want to grow. As an extension developer, I would be happy to see new users and new accidental techies (similar to what @ufundo commented). We should also work on the expectations of what users should expect, and we should make the process easy.

> "some devs might not even make them available on civicrm.org"

I think this is already the case, and since I keep an eye on new extensions, knowing who the devs are, I'd say most people add extensions in the hope that they are discovered and used.  I sometimes nag people to add their extensions to the directory, usually it's just low on their list of priorities, but not something they oppose.

Actually, one problem that made me work on this, is that people ask for a review of their extension, so that it can be installed in-app. However, very few people do [extension reviews](https://lab.civicrm.org/extensions/extension-review-requests/-/issues/). Some have been waiting for 1 year. Personally, I don't feel comfortable doing a review for an extension that has 5 users (and very recently we agreed to [change the criteria for review](https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/1188)). Based on this [discussion on the extension chat](https://chat.civicrm.org/civicrm/pl/azcowqkr4igxmrs39bcfzfgdoc).

So I think the key thing that we do have real feedback on, is that people want their extensions to be available in-app.

However, between the trust model, and the support expectations, based on the comments above, I think we should definitely:

- Have some better information in-app to make an educated decision (clear warning, extension description, support expectations)
- Adapt the terminology on civicrm.org with regards to alpha/beta/stable, so people can leave their extension as beta if they do not want it available in-app (if you're not going to provide support, then don't pretend it's stable? it will definitely break on upgrades at some point)
- Act more quickly when an extension has known problems
- Improve the performance of the extdir module

If there is some agreement towards that, I'm happy to work on that.

**Comment by twomice** [comment]
Very helpful replies, @mlutfy -- thanks!

> People usually report bugs on issue queues. ... Then we can change the stability back to "beta" or "unsupported" and it won't show in-app (and we do so from time to time). We can be more aggressive about updating the status.

Do you believe there's bandwidth available for that additional level of ongoing review? The current limitation of human resources can give us two different problems:

- We currently see a limit in bandwidth to complete reviews; therefore extensions are _not listed in-app_.
- This PR suggests sidestepping that limitation, to get more extensions listed in-app; but if there's still limited bandwidth, are't we likely to see extensions _not being reveiwed for removal_ ?

In other words: Does this approach merely swap one problem for another? And if so, do we prefer the new problem to the old?

**Comment by MegaphoneJon** [comment]
The use case I'm mainly considering here is the new enthusiast.  When I've been hired by amateurs who need professional support, it's common that they've installed a bunch of extensions just to see what they do.

The timeout issue is an annoying one - it leads to long load times when the system checks are run. I don't think having a JSON feed is a blocker, but I do think the timeouts should be addressed first.
What if we put a small proxy in front of the Drupal module?  It could be quite small:
 * Initial requests for the extension list are passed on to Drupal. Results are saved as text files.
 * Future requests for that list return the text file.  *After* the file is loaded, it does a check on the text file's modification date.  If it exceeds the TTL, then the request is sent to Drupal so it's ready for the next person.

We could obviously get more clever - add JSON support, cache the entire extension list and filter within the script, etc.  But the basics should be quick.

@bgm - Is the thought that extension reviews will go away after this?

**Comment by twomice** [comment]
> The use case I'm mainly considering here is the new enthusiast.

Me too. While we as service providers can lock down the extensions or otherwise maintain knowledgeable control, the new enthusiast or non-technical site admin -- I assume -- is both _less_ likely to understand the nuances and make an informed decision, and _more_ likely to have a bad experience if things go wrong with an extension (and also more likely to think poorly of CiviCRM if the extension was listed in-app).

@mlutfy wrote:

> ... I think we should definitely:
> 
> - Have some better information in-app to make an educated decision (clear warning, extension description, support expectations)
> - Adapt the terminology on civicrm.org with regards to alpha/beta/stable, so people can leave their extension as beta if they do not want it available in-app (if you're not going to provide support, then don't pretend it's stable? it will definitely break on upgrades at some point)
> - Act more quickly when an extension has known problems
> - Improve the performance of the extdir module
> 
> If there is some agreement towards that, I'm happy to work on that.

I agree, with some extra emphasis on this one:
> - Have some better information in-app to make an educated decision (clear warning, extension description, support expectations)

I'd like that to say, not just "better information in-app", but "extra steps to even _see_ the unreviewed extensions in-app, such as a 'show unreviewed extensions' checkbox", especially for the sake of those new enthusiasts and non-technical site admins.

**Comment by mlutfy** [comment]
> "Do you believe there's bandwidth available for that additional level of ongoing review?"

There's bandwidth. Personally, I had been reluctant to invest in reviews which, in my opinion, were not always justified. Reviews should be for extensions that have adoption, a good community of devs supporting it, and that we want to promote.

Most of my time in the extension ecosystem is volunteer, but the CT has been investing more time in extensions. For example, Eileen and Coleman have been proactive in sending patches to extensions when they detect core changes that would break extensions. We also moved a few abandoned extensions to Gitlab and consider them "community-maintained" (meaning the CT does not officially maintain them, but a few admins will merge changes and tag releases for maintenance fixes).

> "aren't we likely to see extensions not being reviewed for removal"

I don't think it's a huge concern, but yes, when it happens, we have to be fairly quick on this. I think we should do that regardless of this PR.

> "Is the thought that extension reviews will go away after this?"

They would stay for extensions we recommend. They get bumped to the top of the listing (the sort is by reviewed first, then by usage). There is an expectation that reviewed extensions will provide faster responses to bug reports, and always support the latest stable version of CiviCRM.

> "The timeout issue is an annoying one"

Yes. Fwiw, it does have caching from nginx too, and we can tweak it. That's next on my list.  (and then, composer, because we also need better DX for extensions)

> "non-technical site admin [...] and more likely to have a bad experience if things go wrong with an extension"

Although currently admins have a bad experience because it's clunky to FTP files onto a server.

There is only so much child-proofing we can do, and there is a UX cost to it. People can run things on a sandbox environment, and that's more common nowadays (Bitnami/AWS/etc). If the sandbox crashes, they reset it.

The worst that can go wrong, is a syntax error which stops CiviCRM from loading completely. A bit like when people upgraded to PHP8 and found syntax errors in extensions. Another one was my "report error" extension, where I once forgot to bundle the composer dependencies, and it crashed on load. And it is a reviewed extension (:facepalm:) - a flag I will remove once we fix the "unreviewed extensions do not get upgrade notifications" problem.

So I think that while we can imagine many things going wrong, we have to be pragmatic about what usually goes wrong :)

There are benefits for developers to have their extensions available in-app (even if at the bottom of the listing, at least you can blog about it, recommend it on Stack Exchange). And there are benefits for admin, who will be able to experiment on sandboxes, or yolo on their prod environment. Worst that happens, is that they open a ticket to their hosting provider to ask them to remove the files of an extension, but we have to be pragmatic, that doesn't happen often. We could even pretty easily do some automatic testing to make sure that extensions do not crash-on-install with the latest CiviCRM version, when a new RC is cut.. :thinking:

**Comment by totten** [comment]
> Display all stable extensions (not just reviewed ones)

I'd encourage a UI option to easily switch the feed:

* By default, a site shows extensions from the "Stable (Reviewed)" feed.
* You can click a dropdown to choose alternate feeds, ie
    * "Stable (Reviewed)"
    * "Stable (Unreviewed)"
    * "Development"
* The dropdown simply updates the `ext_repo_url` (as long as the setting is editable).
* This will be "sticky". All your upgrade-alerts will be consistent with this feed-preference. 
* If you want to cherry-pick, you can switch the feed back/forth. (*Turn on the development feed; then install one specific beta-release; then switch back to the stable feed.*)
* You can do this without any major changes/issues in the domain-model.

> Removed "Type" column (was always "module")

+1

> Added "Stability" column (alpha, beta, stable, reviewed)

* Right, you need to show if it's reviewed or not. I guess the minimalist thing would be to add a column for "Reviewed" (`ready`), but that column would feel... somehow... _off_ or _unbalanced_. I like how "Stability" tries to make something more useful.

* But let me offer a critique and alternative.

* __Alternative__: _Put an icon next to the extension-name (no new column)_ (*like a browser "Shield"*)
    * For reviewed extensions, show a positive icon (green/blue/success)  -- perhaps [`check-to-slot` or `list-check` or `ab-check`](https://fontawesome.com/search?q=check&o=r&m=free)). 
    * For unreviewed extensions, show a negative icon (grey/yellow/warning) .
    * You can mouse-over for a 1-sentence explanation. Then click for a link to full explanation.
    * This enables the snap-judgment that we want ("Green Icon - Good") but it doesn't force the reader to make a snap interpretation of the heavier words "Reviewed", "Stable", "Stability". (If they want to understand, then they can drill down on the icon...)

* This circumvents some other bits of confusion that I experience with current rendering. To illustrate, let's fill out a broader example:

        | Extension | Version | Stability |
        | -- | -- | -- |
        | Hello World | 2.0.alpha1 | Reviewed |
        | GBCW | 1.0.beta2 | Beta |
        | IANAL | 1.1 | Stable |
        | LMGTFY | 1.2 | Reviewed |

    * The "Version" already shows "alpha/beta".  So then the reader needs to figure if "Stability" is saying something equivalent or different.

    * Visually, it shows "Hello World" then "2.0.alpha1" then "Reviewed". The reader has to figure if  "Reviewed" is describing "Hello World" or "2.0.alpha1". I bet I'd read it wrong the first few times.

    * If we put the icon by the extension-name, then I think more people will interpret correctly.

    * "Stability" is based on mixing information from different levels (release `develStage` and project `ready`). That's not impossible to do, but it's susceptible to edge-cases.

> "Download" action renamed to "Install" ("download" is a technicality, the actual action is "install")

With either label, I think it's kind of a mess, and it's kind of OK. 

> Removed jsortable, so that the sort order is more predictable (same as on civicrm.org) - I'm on the fence on this, maybe people want to sort alphabetically if they don't know that their browser has a search function?

IIRC, @colemanw had some ambivalence about the existence of jsortable. So probably no one's attached to that implementation. Maybe we remove it -- and then add back something better?

FWIW,  I periodically have to re-check  Civi's installer, which means running through all the CMS management UI's back-to-back. In that scenario...

* I like "Backdrop"'s most. The screen is long, but the filter is good. I type "Civi" and the first hit is exactly what I want.
* I get a bit annoyed in "D7". Yeah, I use browser-search for "Civi". But then it highlights all the descriptions/metadata for Civi-adjacent modules. I have to press `Ctrl-G` several times while carefully watching the highlighter jump around a scrolling screen. It feels like playing Commander Keen, except you don't earn any points.
* WordPress is fine. The default plugin-list is short, so the browser-search is quick/effective.

**Comment by mlutfy** [comment]
> "I'd encourage a UI option to easily switch the feed"

hmm, my knee-jerk reaction is that this would add some unnecessary complexity.

Example:

- I read on Stack Exchange that I need this extension
- first I need to change this setting, maybe it shows some warnings that I don't care about because it's in my way of getting things done
- then I get the listing I expected in the first place.

Then, someone installs an extension manually, because they didn't know about the setting, and they didn't flip the setting, so now they're not getting upgrade alerts. So we would need to implement a separate check. Which is all technically easy enough to do, but I have to question why we would need to do it. What's the real motivation, other than thinking that admins will shoot themselves in the foot. That warning won't stop them, and we're annoying everyone else.

> "Alternative: Put an icon next to the extension-name (no new column) (like a browser "Shield")"

Yep that sounds good. We have the "shield" icon on civicrm.org, but we could also add a warning on civicrm.org as well, to have the same icons / provide the same experience.

I'll update this PR shortly.

**Comment by totten** [comment]
>  We have the "shield" icon... I'll update this PR...

OK, cool 

> > I'd encourage a UI option to easily switch the feed
>
> hmm, my knee-jerk reaction is that this would add some unnecessary complexity.

OK, yeah, the upgrade consideration is important.  After the user chooses to install an unreviewed extension, we shouldn't subject them to any extra work for keeping it up-to-date.

From the POV of "Add New" UX, I don't think it's too complex for the user to have a toggle for reviewed/unreviewed extensions. But that kind of filtering can be in the presentation-layer (Civi PHP or browser-side JS) -- and it's easier to iterate on that bit of UX if it's in presentation-layer.

So, yeah, +1 for same feed for "Stable" releases.

(*For enabling alpha/beta downloads, the calculation would be a bit different -- ie the current feed protocol needs to report one version of each extension. Choosing a different feed is currently straightest path to alphas/betas. Obviously, there's a time/quality trade-off -- if one is willing to spend more time, then more approaches become available.*)

**Comment by ufundo** [comment]
>  "extra steps to even see the unreviewed extensions in-app, such as a 'show unreviewed extensions' checkbox", 

+1 for this UI.

Would this assuage people's concerns about making unreviewed extensions in app installable?

The tickbox could have a health warning popup "installing unstable extensions mighr break your site". If youre a hobby/new user on a dev/throwaway site, that might sound like fun. If you have an extension you own that isnt reviewed because its only relevant to you, you will know the risk you are taking. And it removes barriers for everyone?

I think tickbox filters are a much more natural UI than a feed selector. You could have separate tickboxes for "unreviewed" and "unstable"?

> green shield next to the extension name

+1 for this too

**Comment by mlutfy** [comment]
Here are some changes to address some of the feedback so far:

- Under "Add New", it shows a trophy/warning icon, depending on the review status
- I removed the "status" column, because it's clutter at this point (either "reviewed" or "unreviewed" - I will do something similar on the main ext page)
- Shortened the into help text. I asked non-devs, and they told me the help text was too long, shouldn't be always displayed, highlighted, etc. but that's another topic.

image

When we click "install"

image

image

For the rest of the UX, such as having filters/toggles, I think that would be great to have this in SearchKit eventually. I wouldn't invest too much time into this UI.

**Comment by twomice** [comment]
Great work @mlutfy . I'd really like the "new scary message" to be more scary (or at least, much more visible).  The following looks more appropriate to me -- no changes in language, just in styling. Thoughts?

red
green

File: templates/CRM/Admin/Page/ExtensionDetails.tpl [hunk 1]
```diff
@@ -33,9 +23,19 @@
     </tr>
     {if $extension.develStage}
     <tr>
-      <td class="label">{ts}Development stage{/ts}</td><td>{$extension.develStage|escape}</td>
+      <td class="label">{ts}Stability{/ts}</td><td>{$extension.develStage_formatted|escape}
+        {if $extension.ready == 'ready'}
+          {icon icon="fa-trophy crm-extensions-stage"}Reviewed{/icon} {ts}This extension has been reviewed by the community.{/ts}
+        {elseif $extension.ready == 'not_ready'}
+          {icon icon="fa-warning crm-extensions-stage"}Warning{/icon} {ts}This extensions has not been reviewed by the community.{/ts} {ts}Please proceed with caution when installing a unreviewed extension.{/ts} {docURL page="dev/extensions/lifecycle"}
```
Comment (twomice) [inline templates/CRM/Admin/Page/ExtensionDetails.tpl:30]:
```suggestion
          {icon icon="fa-warning crm-extensions-stage"}Warning{/icon} {ts}This extension has not been reviewed by the community.{/ts} {ts}Please proceed with caution when installing an unreviewed extension.{/ts} {docURL page="dev/extensions/lifecycle"}
```

File: templates/CRM/Admin/Page/Extensions/AddNew.tpl [hunk 2]
```diff
@@ -20,16 +20,21 @@ Depends: CRM/common/enableDisableApi.tpl and CRM/common/jsortable.tpl
...
                 {include file="CRM/Admin/Page/ExtensionDetails.tpl" extension=$row}
             </details>
           </td>
-          <td class="crm-extensions-version">{$row.version|escape}</td>
-          <td class="crm-extensions-description">{$row.type|capitalize}</td>
+          <td class="crm-extensions-version right">{$row.version|escape}
+            {if $row.ready == 'ready'}
+              {icon icon="fa-trophy crm-extensions-stage"}{ts}This extension has been reviewed by the community.{/ts}{/icon}
+            {else}
+              {icon icon="fa-warning crm-extensions-stage"}{ts}This extensions has not been reviewed by the community. Proceed with caution.{/ts}{/icon}
```
Comment (twomice) [inline templates/CRM/Admin/Page/Extensions/AddNew.tpl:35]:
```suggestion
              {icon icon="fa-warning crm-extensions-stage"}{ts}This extension has not been reviewed by the community. Proceed with caution.{/ts}{/icon}
```

**Review by twomice — COMMENTED** [review]
Suggesting minor fixes in grammar.

**Comment by mlutfy** [comment]
Thanks @twomice ! How's this?

image

Was your example with the browser inspector, or using an existing CSS class? I applied the class I found in core. Didn't find anything in Greenwich.

I also like removing the status, which is contradictory/redundant with the text next to it.

And then I added more text, because I figured, if this is going to be scary, it better be useful. Even if the new text is really long, I didn't feel like we were really helping users in taking an educated guess.

**Comment by mlutfy** [comment]
(I removed the trailing "if any" in the warning, it felt redundant)

**Comment by twomice** [comment]
> Was your example with the browser inspector, or using an existing CSS class?

I used class names from core civicrm.css:
- For green, I used `class="extension-installed"`, which I admit is kinda cheating if we think class names mean anything.
- For red, I used `class="crm-error"`, which is not-as-much cheating.

> And then I added more text, because I figured, if this is going to be scary, it better be useful. Even if the new text is really long, I didn't feel like we were really helping users in taking an educated guess.

I liked the shorter text better. My rationale: A short warning on my lawnmower that says "Serious injury may result! Consult owner's manual." is easier to take seriously, compared to a paragraph of text. So I like a short message that says, "Because of XYZ, you should be careful. Read more here ..."

But I claim no special UX knowledge, so of course I'm just offering one opinion.

**Comment by mlutfy** [comment]
I did minor tweaks, but I'll stop here, because with the current state of theming, I don't want to complicate things for Riverlea by adding CSS semantics, even if very tempting (unless @vingle has a strong opinion). Reminds me of #25961.

- I moved the icon next to the version, so that both version numbers look the same (and avoid piling onto the problem above)
- Shortened the message by removing a few words.
- I will ignore the CSS classes for now, because I think it's a wider issue.

On the length of the message itself: I'm the first one to joke about how verbose CiviCRM can be, stating the obvious. However, it seems here like if we insist that this is really something dangerous, then we should do more than displaying images of dead-skulls saying to "stay out". Also, the message is shown at a place and at a moment when it is relevant.

(My coffee machine had a big warning "STOP - read the manual" on the power plug, and it was actually about not calling the shop for support, but rather about calling another number. There was nothing really dangerous about it. So I want to avoid "warning fatigue")

image

**Comment by vingle** [comment]
I prefer the look of @twomice's version, but agree there's no nice utility classes outside of Bootstrap. RiverLee has made Bootstrap's .alert-success/warning/danger/info styles available anywhere, but without that, using `.crm-error` doesn't seem so bad? I'm to not have the green `.extension-installed` bg tho as it might give _too much_ confidence? It's not that each new release is reviewed by 'the community'.

Another thought, the main extensions table has sort icons in the header, which might be useful with such a long list of extensions? Especially with extra column(s) for # of (say) active installs, last release date and/or author? In absence of filters, it might make such a long list a bit easier?

**Comment by mlutfy** [comment]
Sounds good. If Riverlea supports bootstrap alert classes, then I propose adding both `crm-error` and `alert alert-danger` as a transition.

Here is RiverLea with Finsbury variant, with both classes:

image

RiverLea with just `crm-error`:

image

The Island with both classes:

image

and Greenwich, as Allen had suggested:

image

**Comment by mlutfy** [comment]
I added back sorting on the extension title, since I guess since Ctrl+F can be frustrating, as @totten mentioned.

However I don't think we should have a column for the number of installs. I know it's a bit weird to sort on that initially, then not display why, but it feels clutter-y to me, especially since I think we should show numbers like "100+", "500+", not exact numbers (I'll fix that on the extdir at some point).

And of course, that's somewhat temporary until we move it to SearchKit, I hope :)

**Comment by mlutfy** [comment]
I'm sure mergers will be hesitant to merge this, but before one possibly does, are there any major objections that have not been addressed?

**Comment by twomice** [comment]
There will always be room for improvement, but I have no major concerns remaining. Thanks @mlutfy !

**Comment by demeritcowboy** [comment]
Thanks everybody :shipit:
